[SYCL-MLIR]: Zero initialize global variables

etiotto opened the pull request:
This patch initializes tentative definition for global variable to {0}.

Fixes https://github.com/intel/llvm/issues/7355

whitneywhtsang reviewed (commented):
Do you know why the integer element type global in lit test `loop.cpp` and `static.c` are not initialized?

whitneywhtsang commented on polygeist/tools/cgeist/Lib/clang-mlir.cc:
```diff
@@ -2143,8 +2141,34 @@ MLIRASTConsumer::getOrCreateGlobal(const ValueDecl &VD, std::string Prefix,
-  // Initialize the global variable.
-  if (const Expr *Init = Var->getInit()) {
+  // Initialize the global.
+  const VarDecl *InitDecl;
+  const Expr *InitExpr = Var->getAnyInitializer(InitDecl);
```
As `InitDecl` is not used, you can probably use:
```
const Expr *getAnyInitializer() const {
    const VarDecl *D;
    return getAnyInitializer(D);
  }
  ```
```suggestion
  const Expr *InitExpr = Var->getAnyInitializer();
```

victor-eds commented on polygeist/tools/cgeist/Lib/clang-mlir.cc:
```diff
@@ -2143,8 +2141,34 @@ MLIRASTConsumer::getOrCreateGlobal(const ValueDecl &VD, std::string Prefix,
+  const Expr *InitExpr = Var->getAnyInitializer(InitDecl);
+
+  if (!InitExpr) {
+    if (DefKind == VarDecl::TentativeDefinition) {
```
I'd say this is the culprit of the issue @whitneywhtsang mentions. I'm no C expert, but changing the extension of the updated tests from `.c` to `.cpp` makes some global variables `uninitialized`. Maybe we should differ between C and C++ here?

etiotto replied:
I fixed the C++ problems now

etiotto commented:
There should be no difference between C and C++ for global uninitialized variables. The clang/clang++ compilers generate the zero initializer. That is given the following test case: 
```
cat static.c                                                                                                                                                                                      ✔ 
#define N 8

static int B[N]; 
int A[N]; 

int foo() {
  static int bar[N];
  return A[0] + B[2]+bar[0];
}
```
The C compiler generates:
```
./bin/clang -S -emit-llvm static.c; grep global static.ll                                                                                                                                        
@foo.bar = internal global [8 x i32] zeroinitializer, align 16
@A = dso_local global [8 x i32] zeroinitializer, align 16
@B = internal global [8 x i32] zeroinitializer, align 16
```
The C++ compiler generates (static.cpp is a symlink to static.c):
```
./bin/clang++ -S -emit-llvm static.cpp; grep global static.ll                                                                                                                                    
@A = dso_local global [8 x i32] zeroinitializer, align 16
@_ZZ3foovE3bar = internal global [8 x i32] zeroinitializer, align 16
@_ZL1B = internal global [8 x i32] zeroinitializer, align 16
```

I can see that `cgeist` generates the following for a C compilation unit:
```
bin/cgeist static.c --function="*" -S -emit-llvm|grep global                                                                                       
@B = private global [8 x i32] zeroinitializer, align 16
@A = global [8 x i32] zeroinitializer, align 16
@"foo@static@bar" = private global [8 x i32] undef, align 16
```
... and for a C++ compilation unit:
```
bin/cgeist static.cpp --function="*" -S -emit-llvm|grep global                                                                     
@_ZL1B = private global [8 x i32] undef, align 16
@A = global [8 x i32] undef, align 16
@"_Z3foov@static@_ZZ3foovE3bar" = private global [8 x i32] undef, align 16
```

So, currently this PR has addressed some of the issues with default initializations for C compilation units. Remaining issues are:
  - for C,  `@"foo@static@bar" = private global [8 x i32] undef, align 16` is incorrect (shoiuld be sero initialized)
   - for C++ all the globals are incorrectly left uninitialized.

victor-eds commented:
> There should be no difference between C and C++ for global uninitialized variables. The clang/clang++ compilers generate the zero initializer. That is given the following test case:
> 
> ```
> cat static.c                                                                                                                                                                                      ✔ 
> #define N 8
> 
> static int B[N]; 
> int A[N]; 
> 
> int foo() {
>   static int bar[N];
>   return A[0] + B[2]+bar[0];
> }
> ```
> 
> The C compiler generates:
> 
> ```
> ./bin/clang -S -emit-llvm static.c; grep global static.ll                                                                                                                                        
> @foo.bar = internal global [8 x i32] zeroinitializer, align 16
> @A = dso_local global [8 x i32] zeroinitializer, align 16
> @B = internal global [8 x i32] zeroinitializer, align 16
> ```
> 
> The C++ compiler generates (static.cpp is a symlink to static.c):
> 
> ```
> ./bin/clang++ -S -emit-llvm static.cpp; grep global static.ll                                                                                                                                    
> @A = dso_local global [8 x i32] zeroinitializer, align 16
> @_ZZ3foovE3bar = internal global [8 x i32] zeroinitializer, align 16
> @_ZL1B = internal global [8 x i32] zeroinitializer, align 16
> ```
> 
> I can see that `cgeist` generates the following for a C compilation unit:
> 
> ```
> bin/cgeist static.c --function="*" -S -emit-llvm|grep global                                                                                       
> @B = private global [8 x i32] zeroinitializer, align 16
> @A = global [8 x i32] zeroinitializer, align 16
> @"foo@static@bar" = private global [8 x i32] undef, align 16
> ```
> 
> ... and for a C++ compilation unit:
> 
> ```
> bin/cgeist static.cpp --function="*" -S -emit-llvm|grep global                                                                     
> @_ZL1B = private global [8 x i32] undef, align 16
> @A = global [8 x i32] undef, align 16
> @"_Z3foov@static@_ZZ3foovE3bar" = private global [8 x i32] undef, align 16
> ```
> 
> So, currently this PR has addressed some of the issues with default initializations for C compilation units. Remaining issues are:
> 
>     * for C,  `@"foo@static@bar" = private global [8 x i32] undef, align 16` is incorrect (should be zero initialized)
> 
>     * for C++ all the globals are incorrectly left uninitialized.



> There should be no difference between C and C++ for global uninitialized variables. The clang/clang++ compilers generate the zero initializer. That is given the following test case:
> 
> ```
> cat static.c                                                                                                                                                                                      ✔ 
> #define N 8
> 
> static int B[N]; 
> int A[N]; 
> 
> int foo() {
>   static int bar[N];
>   return A[0] + B[2]+bar[0];
> }
> ```
> 
> The C compiler generates:
> 
> ```
> ./bin/clang -S -emit-llvm static.c; grep global static.ll                                                                                                                                        
> @foo.bar = internal global [8 x i32] zeroinitializer, align 16
> @A = dso_local global [8 x i32] zeroinitializer, align 16
> @B = internal global [8 x i32] zeroinitializer, align 16
> ```
> 
> The C++ compiler generates (static.cpp is a symlink to static.c):
> 
> ```
> ./bin/clang++ -S -emit-llvm static.cpp; grep global static.ll                                                                                                                                    
> @A = dso_local global [8 x i32] zeroinitializer, align 16
> @_ZZ3foovE3bar = internal global [8 x i32] zeroinitializer, align 16
> @_ZL1B = internal global [8 x i32] zeroinitializer, align 16
> ```
> 
> I can see that `cgeist` generates the following for a C compilation unit:
> 
> ```
> bin/cgeist static.c --function="*" -S -emit-llvm|grep global                                                                                       
> @B = private global [8 x i32] zeroinitializer, align 16
> @A = global [8 x i32] zeroinitializer, align 16
> @"foo@static@bar" = private global [8 x i32] undef, align 16
> ```
> 
> ... and for a C++ compilation unit:
> 
> ```
> bin/cgeist static.cpp --function="*" -S -emit-llvm|grep global                                                                     
> @_ZL1B = private global [8 x i32] undef, align 16
> @A = global [8 x i32] undef, align 16
> @"_Z3foov@static@_ZZ3foovE3bar" = private global [8 x i32] undef, align 16
> ```
> 
> So, currently this PR has addressed some of the issues with default initializations for C compilation units. Remaining issues are:
> 
>     * for C,  `@"foo@static@bar" = private global [8 x i32] undef, align 16` is incorrect (should be zero initialized)
> 
>     * for C++ all the globals are incorrectly left uninitialized.

Yes, that's what I meant. I'm not saying that `clang`'s output differs between C and C++, but that ours does

whitneywhtsang commented on polygeist/tools/cgeist/Test/Verification/static.c:
```diff
@@ -6,8 +6,8 @@ int foo() {
 
-// CHECK:   memref.global "private" @"foo@static@bar" : memref<8xi32> = uninitialized {alignment = 16 : i64}
-// CHECK:   func @foo() -> i32 attributes {llvm.linkage = #llvm.linkage<external>} {
+// CHECK-DAG:   memref.global "private" @"foo@static@bar" : memref<8xi32> = dense<0> {alignment = 16 : i64}
```
CHECK-DAG is not useful by its own.

etiotto replied:
done
